### PR TITLE
kw2xrf: adaption to netdev2

### PR DESCRIPTION
**Pull request by jfischer-phytec-iot**
This PR adapts the kw2xrf driver to netdev2 and adds new timer and test modes functionalities.
Also support for CR20A devices will be tested. WIP, not ready for review

depends on ~~#5485~~, ~~#5487~~

### CODE REVIEW

**Comment by miri64** [comment]
Needs rebase.

**Comment by miri64** [comment]
Needs rebase and adaption to https://github.com/RIOT-OS/RIOT/pull/4871.

**Comment by miri64** [comment]
Sorry, another adaption needed: #5495. :/

File: cpu/cortexm_common/vectors_cortexm.c [hunk 1]
```diff
@@ -205,8 +205,8 @@ __attribute__((used)) void hard_fault_handler(uint32_t* sp, uint32_t corrupted,
     uint32_t dfsr  = SCB->DFSR;
     uint32_t afsr  = SCB->AFSR;
 #endif
-    uint32_t pc;
-    uint32_t* orig_sp;
+    uint32_t pc = 0;
+    uint32_t* orig_sp = NULL;
```
Comment (miri64) [inline cpu/cortexm_common/vectors_cortexm.c]: Seems like an unrelated fix.

**Comment by miri64** [comment]
There seems to be more than just moving to netdev2 here. Is it maybe possible to split this up in several PRs?

**Comment by miri64** [comment]
(I know this is reflected in the description, but not in the title).

File: cpu/cortexm_common/vectors_cortexm.c (hunk 1 above)
Comment (jfischer-phytec-iot) [inline cpu/cortexm_common/vectors_cortexm.c]: It will/should be removed after rebase.

**Comment by smlng** [comment]
I'd like to test this one. Could you @jfischer-phytec-iot please rebase, and maybe adapt #5834. 

Why is #5487 needed?

**Comment by jfischer-phytec-iot** [comment]
>Why is #5487 needed?

to display actual CCA-Mode ...

File: cpu/cortexm_common/vectors_cortexm.c (hunk 1 above)
Comment (jfischer-phytec-iot) [inline cpu/cortexm_common/vectors_cortexm.c]: ...when #5727 was merged

**Comment by jfischer-phytec-iot** [comment]
@smlng rebased on #5487 and master, I have dropped #5834. The new driver may have some issues, e.g. state tracing, that appeared with the openthread package.

**Comment by smlng** [comment]
@jfischer-phytec-iot: I'm still testing this one. However, you somewhat changed the hardware adress generation, its basically reads backwards from what it was before. Any particular reason to do that? I labeled my nodes with their link-local address - changing the hardware address means changing the lladdr, too - hence, I would have to relabel all (~30) my nodes, which would be a little bit _inconvenient_ 😬

**Comment by smlng** [comment]
I looked into the responsible code, but it looks the same as before. But still compiling with current master I get a different hwaddr (and lladdr) compared to this PR? Weird ...

**Comment by jfischer-phytec-iot** [comment]
>However, you somewhat changed the hardware adress generation, its basically reads backwards from what it was before. Any particular reason to do that?

Yes, i changed it :smile: , sorry. Now, least significant bytes of LL references to the the least significant bytes of CPUID.

**Comment by smlng** [comment]
So you want to keep it that way, and I have to relabel my nodes? Well, then NACK 😁 ... No, I'm fine with it - I just have to remember it to not get confused, when this one is merged.

**Comment by smlng** [comment]
Ah, and could you please fix this conflict in `auto_init_kw2xr.c` in the meantime.

**Comment by smlng** [comment]
I ran some tests on this PR, I tried to ping a `samr21-xpro`, both running latest RIOT and `gnrc_networking` example; and I also tried to ping a RPi with Openlabs transceiver. Additionally, I had a network sniffer running on another RPi, these are the results:

* ping6 `pba-d-01-kw2x` (netdev2) -> `samr21-xpro`: I see the ping request of kw2x on the sniffer and a subsequent ACK. Afterwards I see 4 ping replies for each ping request (3 retrans) by the `samr21-xpro`, but no ACKs by `kw2x` though the ACKreq flag is set by `samr21-xpro`. `kw2x` reports 100% packet loss, though sniffer _sees_ all packets.

* ping6 `pba-d-01-kw2x` (netdev2) -> RPi:  I only see ping requests no reply by the RPi, result is again 100% packet loss. This might be due to the fact that linux drops all packet with INTRA-PAN flag `false`  even though src and dst PAN are equal.

@jfischer-phytec-iot please set INTRA-PAN flag true if src and dst PAN are equal, see also #5684 and #5685. The latter might fix this already.

**Comment by smlng** [comment]
I applied #5685, now UDP send and recv works in both directions with `samr21-xpro`, pings still don't work same behavior as described above. Same for UDP with RPi: works in both directions. Ping6 does not work, from `kw2x` to RPi I see requests and replies, but `kw2x` still reports 100% packet loss. Ping from RPi to `kw2x` I see only ping requests on the sniffer but no reply by `kw2x`.

**Comment by jfischer-phytec-iot** [comment]
@smlng Can you please repeat the ping6 test without 1f4fb5d ?

**Comment by smlng** [comment]
@jfischer-phytec-iot without debugging ping works with `samr21-xpro` and RPi. - nice!

What do you have on your todo list for this one? Otherwise I would give an ACK, so we can move on.

**Comment by jfischer-phytec-iot** [comment]
My TODO list:
- test with the NXP hardware, FRDM-K64F and FRDM-CR20A (equal transceiver as inside kw2x)
- improve code quality
- replacing weird debug outputs with "high speed" hardware tracing (toggle GPIO and decode with usb connected LA and sigrok, kind of https://github.com/jfischer-phytec-iot/rusbdtrace, I use it for the usbdevice stack).

**Comment by miri64** [comment]
Can we get this merged for this release? I know @jfischer-phytec-iot wants to work a little bit more on this PR, but @smlng [really wants to have this in ASAP](https://github.com/RIOT-OS/RIOT/issues/4822#issuecomment-254464710). Maybe basic porting can be done in this PR while anything beyond that can be provided in follow-up PRs?

File: drivers/include/kw2xrf.h [hunk 2]
```diff
@@ -69,90 +62,112 @@ extern "C" {
 #define KW2XRF_DEFAULT_CHANNEL (DEFAULT_CHANNEL)
 #endif
 #ifndef KW2XRF_DEFAULT_CHANNEL
-#define KW2XRF_DEFAULT_CHANNEL        (26U)
+#define KW2XRF_DEFAULT_CHANNEL          (26U)
```
Comment (miri64) [inline drivers/include/kw2xrf.h]: Might need adaptation after #5977 was merged

**Comment by miri64** [comment]
Feature freeze -> postponed

**Comment by haukepetersen** [comment]
rebase needed

**Comment by smlng** [comment]
@jfischer-phytec-iot can we get this merged soon? Might be easier to do if not depending on #5487, please adapt/rebase accordingly. If you don't have time for this right now, I'm willing to overtake this PR and make necessary changes, if you don't mind?!

**Comment by jfischer-phytec-iot** [comment]
will be done today

**Comment by smlng** [comment]
Is there consensus on this `iwpan` shell command stuff? If not, I would recommend to remove it; and even if so, I'd like to see this in a separate PR.

**Comment by miri64** [comment]
@smlng: #5750

**Comment by smlng** [comment]
@miri64 I still don't see that this PR must be based on (or wait for) #5750, I'd rather have this rebased without this dependency and merged quickly.

@jfischer-phytec-iot is it possible to rebase without #5750?

File: drivers/kw2xrf/include/kw2xrf_reg.h [hunk 3]
```diff
@@ -124,13 +142,16 @@ enum mkw2xdrf_dregister {
 #define MKW2XDM_IRQSTS3_TMR3IRQ                     (1 << 2)
 #define MKW2XDM_IRQSTS3_TMR2IRQ                     (1 << 1)
 #define MKW2XDM_IRQSTS3_TMR1IRQ                     (1 << 0)
+#define MKW2XDM_IRQSTS3_TMR_IRQ_MASK                0xfu
```
Comment (jfischer-phytec-iot) [inline drivers/kw2xrf/include/kw2xrf_reg.h]: no, is not necessary

**Comment by jfischer-phytec-iot** [comment]
>@jfischer-phytec-iot is it possible to rebase without #5750?

I will try it.

**Comment by smlng** [comment]
ping

**Comment by miri64** [comment]
ping

**Comment by PeterKietzmann** [comment]
@jfischer-phytec-iot we'd like to remove the netdev1 interface after the release. Any chance you prepare this PR until middle of next week so @smlng can take over the netdev2 adoption part of this pr?

File: drivers/kw2xrf/kw2xrf_netdev.c [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+	case NETOPT_STATE_OFF:
+            /* TODO: Replace with powerdown (set reset input low) */
+            kw2xrf_set_power_mode(dev, KW2XRF_HIBERNATE);
+        default:
+            return -ENOTSUP;
+    }
+    return sizeof(netopt_state_t);
+}
+
+netopt_state_t _get_state(kw2xrf_t *dev)
```
Comment (smlng) [inline drivers/kw2xrf/kw2xrf_netdev.c]: should be static, otherwise name clash with other drivers, e.g. at86rf233, see Jenkins log [here](https://riot-ci.inet.haw-hamburg.de/job/RIOT/job/PR-5469/3/artifact/error_iotlab-a8-m3_linux_driver_tests.log)

Comment (jfischer-phytec-iot) [inline drivers/kw2xrf/kw2xrf_netdev.c]: ok

**Comment by PeterKietzmann** [comment]
@jfischer-phytec-iot we would really much like to have this adoption a a part of the 2016.01 release so we can remove the obsolete netdev1 API. Is there any change that you squash all your commits until the afternoon???

**Comment by jfischer-phytec-iot** [comment]
@PeterKietzmann I will not make as long as checks are red

**Comment by smlng** [comment]
needs rebase again, that's why CI does not build

**Comment by jfischer-phytec-iot** [comment]
I give up, that was the last drop:
drivers/kw2xrf/kw2xrf_intern.c:35: style (unreadVariable): Variable 'reg' is assigned a value that is never used.

@PeterKietzmann @smlng If anyone makes PR against this branch, I can merge and squash it, but I do not have time for this today.

File: drivers/kw2xrf/kw2xrf_intern.c [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+void kw2xrf_disable_interrupts(kw2xrf_t *dev)
+{
+    /* Clear and disable all interrupts */
+    kw2xrf_write_dreg(dev, MKW2XDM_PHY_CTRL2, 0xff);
+    int reg = kw2xrf_read_dreg(dev, MKW2XDM_PHY_CTRL3);
+    reg |= MKW2XDM_PHY_CTRL3_WAKE_MSK | MKW2XDM_PHY_CTRL3_PB_ERR_MSK;
```
Comment (miri64) [inline drivers/kw2xrf/kw2xrf_intern.c]: The `cppcheck` actually is right. This variable is set, but never read again.

Comment (smlng) [inline drivers/kw2xrf/kw2xrf_intern.c]: mhm, maybe a call is missing, in the current driver (`kw2xrf.c`) I found:

```
        /* Mask all possible interrupts */
        reg = kw2xrf_read_dreg(MKW2XDM_PHY_CTRL3);
        reg |= MKW2XDM_PHY_CTRL3_WAKE_MSK;
        kw2xrf_write_dreg(MKW2XDM_PHY_CTRL3, reg);

        kw2xrf_write_dreg(MKW2XDM_IRQSTS1, 0x7f);
        kw2xrf_write_dreg(MKW2XDM_IRQSTS2, 0x03);
        kw2xrf_write_dreg(MKW2XDM_IRQSTS3, 0xff);
```

so `reg` is written back to the register, that is missing here?!

Comment (smlng) [inline drivers/kw2xrf/kw2xrf_intern.c]: ah the following line 

```
kw2xrf_write_dreg(dev, MKW2XDM_PHY_CTRL3, 0x03);
```

here actually `0x03` should be `reg` because `MKW2XDM_PHY_CTRL3_WAKE_MSK | MKW2XDM_PHY_CTRL3_PB_ERR_MSK = 0x03` according to the defines.

so fixing this by either remove `reg` or better as `0x03` is some magic number with
```
kw2xrf_write_dreg(dev, MKW2XDM_PHY_CTRL3, reg);
```

should do

Comment (jfischer-phytec-iot) [inline drivers/kw2xrf/kw2xrf_intern.c]: will check squashed branch against backup

**Comment by smlng** [comment]
closing, superseded by #6453

**Comment by miri64** [comment]
Wanted to try something
